File: leetcode/check-if-two-string-arrays-are-equivalent.py
# 1662. Check If Two String Arrays are Equivalent
# 🟢 Easy
#
# https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent/
#
# Tags: Array - String

import timeit
from collections.abc import Generator
from itertools import chain, zip_longest
from typing import List


# Convert the arrays to strings using join() then compare them.
#
# Time complexity: O(m+n) - The sum of the length of each string.
# Space complexity: O(m+n) - The strings that we cast the arrays to.
#
# Runtime: 61 ms, faster than 46.31%
# Memory Usage: 13.9 MB, less than 75.82%
class Naive:
    def arrayStringsAreEqual(self, word1: List[str], word2: List[str]) -> bool:
        return "".join(word1) == "".join(word2)


# We can improve on the previous solution by defining a generator
# function that produces the characters of a given List[str] in order.
#
# Time complexity: O(min(m, n)) - The algorithm will run at most the
# length of the shorter input counting characters.
# Space complexity: O(1) - No extra space is used, the generator does
# not store all values in memory but only uses a constant amount of
# memory.
#
# Runtime: 62 ms, faster than 43.12%
# Memory Usage: 13.9 MB, less than 33.11%
class UseGenerator:
    def arrayStringsAreEqual(self, word1: List[str], word2: List[str]) -> bool:
        # Define a helper function to help iterate over the inputs.
        def getChars(arr: List[str]) -> Generator[str, None]:
            for chars in arr:
                for char in chars:
                    yield char
            # Mark the end of the stream.
            yield None

        return all(a == b for a, b in zip(getChars(word1), getChars(word2)))


# We can improve even further using one of the functions in the
# itertools package instead of defining our own.
#
# Time complexity: O(max(m, n)) - The algorithm will run at most the
# length of the longest input counting characters.
# Space complexity: O(1) - No extra space is used, the generator does
# not store all values in memory but only uses a constant amount of
# memory.
#
# Runtime: 57 ms, faster than 58.20%
# Memory Usage: 13.9 MB, less than 33.11%
class BuiltInFn:
    def arrayStringsAreEqual(self, word1: List[str], word2: List[str]) -> bool:
        return all(
            a == b
            for a, b in zip_longest(
                chain.from_iterable(word1), chain.from_iterable(word2)
            )
        )


def test():
    executors = [
        Naive,
        UseGenerator,
        BuiltInFn,
    ]
    tests = [
        [["ab", "c"], ["a", "bc"], True],
        [["a", "cb"], ["a", "c"], False],
        [["a", "cb"], ["ab", "c"], False],
        [["abc", "d", "defg"], ["abcddefg"], True],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.arrayStringsAreEqual(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/check-if-two-string-arrays-are-equivalent.rs
// 1662. Check If Two String Arrays are Equivalent
// 🟢 Easy
//
// https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent/
//
// Tags: Array - String

use itertools::EitherOrBoth::{Both, Left, Right};
use itertools::Itertools;

struct Solution;
impl Solution {
    #[allow(dead_code)]
    /// This solution does not run on LeetCode.
    pub fn array_strings_are_equal(word1: Vec<String>, word2: Vec<String>) -> bool {
        word1
            .iter()
            .flat_map(|s| s.chars())
            .zip_longest(word2.iter().flat_map(|s| s.chars()))
            .all(|x| match x {
                Both(a, b) => a == b,
                Left(_) | Right(_) => false,
            })
    }

    /// Create a flattened iterator for each of the inputs, iterate over the characters generated
    /// by calling next on both iterators until we don't have more characters or the values
    /// generated don't match.
    ///
    /// Time complexity: O(min(m,n)) - Where m, n is the number of characters in all the Strings in
    /// word1, word2. The algorithm checks characters against each other until if finds a pair that
    /// is not equal or it runs out of characters in one of the inputs.
    /// Space complexity: O(1) - We use iterators to avoid allocating extra memory.
    ///
    /// Runtime 1 ms Beats 89.29%
    /// Memory 1.96 MB Beats 100%
    pub fn array_strings_are_equal_2(word1: Vec<String>, word2: Vec<String>) -> bool {
        let mut it1 = word1.iter().flat_map(|s| s.chars());
        let mut it2 = word2.iter().flat_map(|s| s.chars());
        loop {
            match (it1.next(), it2.next()) {
                (None, None) => return true,
                (None, Some(_)) | (Some(_), None) => return false,
                (Some(a), Some(b)) => {
                    if a != b {
                        return false;
                    }
                }
            }
        }
    }
}

// Tests.
fn main() {
    let tests = [
        (vec!["ab", "c"], vec!["a", "bc"], true),
        (vec!["a", "cb"], vec!["ab", "c"], false),
        (vec!["abc", "d", "defg"], vec!["abcddef"], false),
        (vec!["abc", "d", "defg"], vec!["abcddefg"], true),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::array_strings_are_equal_2(
            t.0.iter().map(|x| x.to_string()).collect::<Vec<_>>(),
            t.1.iter().map(|x| x.to_string()).collect::<Vec<_>>(),
        );
        if res == t.2 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!("");
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
